File: backend/app/api/routes/markets.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import CurrentUser, get_current_user
from app.db.session import get_db
from app.services.market_data import market_bars, market_detail

router = APIRouter(prefix="/markets", tags=["markets"])


@router.get("/{market_id}")
async def detail(
    market_id: str,
    _: CurrentUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> dict:
    item = await market_detail(session, market_id)
    if item is None:
        raise HTTPException(status_code=404, detail="Market not found")
    return item


@router.get("/{market_id}/bars")
async def bars(
    market_id: str,
    range: str = Query("7d", pattern="^(24h|7d|30d|all)$"),
    resolution: str = Query("hour1", pattern="^(min15|hour1|hour4|day1)$"),
    _: CurrentUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> dict:
    return await market_bars(session, market_id, range_name=range, resolution=resolution)

File: mobile/src/api/markets.ts
import { apiFetch } from "./client";
import { toQueryString } from "./queryString";
import type {
  MarketBarsResponse,
  MarketDetail,
  MarketRefreshResponse,
} from "./types";

export interface MarketBarsParams {
  range?: "24h" | "7d" | "30d" | "all";
  resolution?: "min15" | "hour1" | "hour4" | "day1";
}

export const marketKeys = {
  detail: (marketId: string) => ["markets", marketId] as const,
  bars: (marketId: string, params: MarketBarsParams) =>
    ["markets", marketId, "bars", params] as const,
};

export function getMarket(marketId: string) {
  return apiFetch<MarketDetail>(`/markets/${encodeURIComponent(marketId)}`);
}

export function getMarketBars(
  marketId: string,
  params: MarketBarsParams = { range: "7d", resolution: "hour1" },
) {
  return apiFetch<MarketBarsResponse>(
    `/markets/${encodeURIComponent(marketId)}/bars${toQueryString(params)}`,
  );
}

export function refreshMarket(marketId: string) {
  return apiFetch<MarketRefreshResponse>(
    `/markets/${encodeURIComponent(marketId)}/refresh`,
    {
      method: "POST",
    },
  );
}
